packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isFalse, isUndefined, isNull } from '@lwc/shared';
+import { VM, scheduleRehydration } from './vm';
+import { markComponentAsDirty, ComponentConstructor } from './component';
+import { Template } from './template';
+import { StylesheetFactory } from './stylesheet';
+
+const swappedTemplateMap = new WeakMap<Template, Template>();
+const swappedComponentMap = new WeakMap<ComponentConstructor, ComponentConstructor>();
+const swappedStyleMap = new WeakMap<StylesheetFactory, StylesheetFactory>();
+
+const activeTemplates = new WeakMap<Template, Set<VM>>();
+const activeComponents = new WeakMap<ComponentConstructor, Set<VM>>();
+const activeStyles = new WeakMap<StylesheetFactory, Set<VM>>();
+
+function rehydrateHotTemplate(tpl: Template): boolean {
+    const list = activeTemplates.get(tpl);
+    list?.forEach((vm) => {
+        if (isFalse(vm.isDirty)) {
+            // forcing the vm to rehydrate in the next tick
+            markComponentAsDirty(vm);
+            scheduleRehydration(vm);
+        }
+    });
+    return true;
+}
+
+function rehydrateHotStyle(style: StylesheetFactory): boolean {
+    const list = activeStyles.get(style);
+    list?.forEach((vm) => {
+        // hacky way to force the styles to get recomputed
+        // by replacing the value of old template, which is used
+        // during the rendering process. If the template returned
+        // by render() is different from the previous stored template
+        // the styles will be reset, along with the content of the
+        // shadow, this way we can guarantee that the styles will be
+        // recalculated, and applied.
+        vm.cmpTemplate = () => [];
+        if (isFalse(vm.isDirty)) {
+            // forcing the vm to rehydrate in the next tick
+            markComponentAsDirty(vm);
+            scheduleRehydration(vm);
+        }
+    });
+    return true;
+}
+
+function rehydrateHotComponent(Ctor: ComponentConstructor): boolean {
+    const list = activeComponents.get(Ctor);
+    let canRefreshAllInstances = true;
+    list?.forEach((vm) => {
+        const { owner } = vm;
```

Comment:
another point to clear up about this PR is whether or not we should clean up the active VMs associated to the provided Ctor. IMO, this is not a problem... yes, you have things in the weakmap that are anchoring those VMs to old versions of Ctor, and probably a new one too after the rehydration... but that should not represent a problem at all... this is a dev mode feature anyway. But if you feel that we should do the clean up, I can do that, it is simpler.

Reply:
I do think we should cleanup when possible. The 6 weakmaps makes me nervous about memory leaks.
The three weak maps for the swaps allow circular references. The other three weakmaps create a link between the swaps and the vms. Now, the VMs themselves have a strong reference to the template and the component constructors.

Reply:
doing it.